rpcserver.go

```diff
@@ -767,13 +768,89 @@ func (r *rpcServer) SendCoins(ctx context.Context,
 		return nil, err
 	}
 
-	rpcsLog.Infof("[sendcoins] addr=%v, amt=%v, sat/kw=%v", in.Addr,
-		btcutil.Amount(in.Amount), int64(feePerKw))
+	rpcsLog.Infof("[sendcoins] addr=%v, amt=%v, sat/kw=%v, sweep_all=%v",
+		in.Addr, btcutil.Amount(in.Amount), int64(feePerKw),
+		in.SendAll)
 
-	paymentMap := map[string]int64{in.Addr: in.Amount}
-	txid, err := r.sendCoinsOnChain(paymentMap, feePerKw)
-	if err != nil {
-		return nil, err
+	var txid *chainhash.Hash
+
+	wallet := r.server.cc.wallet
+
+	// If the send all flag is active, then we'll attempt to sweep all the
+	// coins in the wallet in a single transaction (if possible),
+	// otherwise, we'll respect the amount, and attempt a regular 2-output
+	// send.
+	if in.SendAll {
+		// As we need to sweep ALL the outputs in the wallet, we'll
+		// construct an ephemeral sweeper instance that will target out
+		// desired sweep address.
+		sweepCfg := sweep.UtxoSweeperConfig{
+			GenSweepScript: func() ([]byte, error) {
+				addr, err := btcutil.DecodeAddress(
+					in.Addr, activeNetParams.Params,
+				)
+				if err != nil {
+					return nil, err
+				}
+
+				return txscript.PayToAddrScript(addr)
+			},
+			FeeEstimator: r.server.cc.feeEstimator,
+			Signer:       r.server.cc.signer,
+		}
+		_, bestHeight, err := r.server.cc.chainIO.GetBestBlock()
+		if err != nil {
+			return nil, err
+		}
+
+		// With the sweeper instance created, we can now generate a
+		// transaction that will sweep ALL outputs from the wallet in a
+		// single transaction. This will be generated in a concurrent
+		// safe manner, so no need to worry about locking.
+		sweepTxPkg, err := sweep.CraftSweepAllTx(
+			sweep.New(&sweepCfg), feePerKw, uint32(bestHeight), wallet,
```

Comment:
They're effectively the same method. The only difference would be getting the delivery address and fee rate at the call site or not. Or do you mean that we shouldn't need to initialize a temp sweeper instance (notice we don't start any of its goroutines) for this method?

Reply:
Yes indeed. We don't start any goroutines, but `sweeper` is designed to be an active component. That we can just create a temp instance, relies on knowledge of the internal implementation of sweeper. Knowing that there are no side effects to this. I think this can cause confusion for people looking at it in the future.

Reply:
Alright, in the latest diff I've modified this to use `createSweepTx` directly. In the future, we may also want to export this method and remove the existing `sweeper.CreateSweepTx` method all together.